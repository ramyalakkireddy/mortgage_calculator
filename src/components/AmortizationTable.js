import React from 'react';
import { makeStyles, withStyles } from '@material-ui/core/styles';
import {
  Table, Grid, Typography
} from '@material-ui/core';
import TableBody from '@material-ui/core/TableBody';
import TableCell from '@material-ui/core/TableCell';
import TableContainer from '@material-ui/core/TableContainer';
import TableHead from '@material-ui/core/TableHead';
import TableRow from '@material-ui/core/TableRow';
import TablePagination from '@material-ui/core/TablePagination';

const StyledTableCell = withStyles((theme) => ({
  head: {
    backgroundColor: 'dimgrey',
    fontFamily: 'serif',
    color: theme.palette.common.white,
  },
  body: {
    fontSize: 14,
  },
}))(TableCell);

const StyledTableRow = withStyles((theme) => ({
  root: {
    '&:nth-of-type(odd)': {
      backgroundColor: theme.palette.action.hover,
    },
  },
}))(TableRow);

const formatter = (amount) => {
  return (
    new Intl.NumberFormat('en-CA',
      { style: 'currency', currency: 'CAD' }
    ).format(amount)
  );
}

const useStyles = makeStyles(theme => ({
  container: {
    minHeight: 'auto'
  },
  text: {
    margin: '20px 0',
    fontFamily: 'serif',
    fontWeight: '600',
  },
  table: {
    minWidth: 650,
    overflow: 'hidden'
  },
  tableRow: {
    fontSize:  theme.typography.pxToRem(16)
  }
}));

const columns = [
  { id: 'year', label: 'Year' },
  { id: 'interest', label: 'Interest Paid' },
  { id: 'principal', label: 'Principal Paid' },
  { id: 'balance', label: 'Balance' }
];

export default function AmortizationTable(props) {
  const classes = useStyles();
  const { amortization } = props;

  const [page, setPage] = React.useState(0);
  const [rowsPerPage, setRowsPerPage] = React.useState(10);

  const handleChangePage = (event, newPage) => {
    setPage(newPage);
  };

  const handleChangeRowsPerPage = event => {
    setRowsPerPage(+event.target.value);
    setPage(0);
  };

  return (
      <Grid container direction="row">
        <Grid item xs={12}>
          <Typography variant="h5" className={classes.text}>Amortization Table:</Typography>
          <TableContainer className={classes.tableContainer}>
            <Table className={classes.table} aria-label="simple table">
            <TableHead className={classes.header}>
                <TableRow>
                  {columns.map(column => (
                    <StyledTableCell
                      className={classes.tableRow}
                      key={column.id}
                      align="left"
                    >{column.label}
                    </StyledTableCell>
                  ))}
                </TableRow>
              </TableHead>
                <TableBody>
                  {amortization.length > 0 && amortization
                    .slice(page * rowsPerPage, page * rowsPerPage + rowsPerPage)
                      .map((data, index) => {
                        return (
                          <StyledTableRow hover key={data.year}>
                            <StyledTableCell>{data.year}</StyledTableCell>
                            <StyledTableCell>{formatter(data.interestN)}</StyledTableCell>
                            <StyledTableCell>{formatter(data.principalN)}</StyledTableCell>
                            <StyledTableCell>{formatter(data.balance)}</StyledTableCell>
                          </StyledTableRow>
                        );
                      })}
                </TableBody>
            </Table>
          </TableContainer>
          <TablePagination
            rowsPerPageOptions={[10, 25, 50, 100]}
            component="div"
            count={amortization.length}
            rowsPerPage={rowsPerPage}
            page={page}
            onChangePage={handleChangePage}
            onChangeRowsPerPage={handleChangeRowsPerPage}
          />
        </Grid>
      </Grid>
  );
};
