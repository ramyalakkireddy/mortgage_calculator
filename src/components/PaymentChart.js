import React from 'react';
import { makeStyles } from '@material-ui/core/styles';
import { Button, Paper, Dialog, DialogActions, DialogContent
} from '@material-ui/core';
import {
  Chart,
  BarSeries,
  Title,
  ArgumentAxis,
  ValueAxis, Tooltip
} from '@devexpress/dx-react-chart-material-ui';
import { Animation, EventTracker } from '@devexpress/dx-react-chart';

const useStyles = makeStyles(theme => ({
  root: {
    overflowY: 'auto',
    paddingTop: '0px !important'
  }
}));

export default function PaymentChart(props) {
  const classes = useStyles();
  const { amortization, onClose, open } = props;

  const handleClose = () => {
    onClose();
  };

  return (
    <Dialog
      onClose={handleClose}
      aria-labelledby="simple-dialog-title"
      open={open}
      fullWidth
      maxWidth={false}
    >
      <DialogContent className={classes.root}>
        <Paper elevation={3}>
          <Chart
            data={amortization}
          >
            <ArgumentAxis />
            <ValueAxis />
            <BarSeries
              valueField="balance"
              argumentField="year"
            />
            <Title text="Payment Graph" />
            <EventTracker />
            <Tooltip />
            <Animation />
          </Chart>
        </Paper>
      </DialogContent>
      <DialogActions>
        <Button onClick={handleClose} variant="contained" color="primary">
          Close
        </Button>
      </DialogActions>
    </Dialog>
  );
}
