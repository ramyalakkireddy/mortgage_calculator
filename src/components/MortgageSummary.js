import React from 'react';
import { Grid, Typography, FormLabel, Button, Divider, Card, CardContent } from '@material-ui/core';
import { makeStyles } from '@material-ui/core/styles';
import BarChart from '@material-ui/icons/BarChart';
import Edit from '@material-ui/icons/EditTwoTone';
import AmortizationTable from './AmortizationTable';
import PaymentChart from './PaymentChart';

const useStyles = makeStyles(theme => ({
  text: {
    fontFamily: 'serif',
    fontWeight: '600',
  },
  title: {
    marginTop: '30px',
    fontFamily: 'serif',
  },
  card: {
    height: '100%',
    backgroundColor: 'lightseagreen',
    color: '#ffffff',
  },
  amount: {
    fontFamily: 'serif',
  },
  grid: {
    textAlign: 'center',
  },
  formLabel: {
    color: '#000000',
    fontSize: '1.15em',
    fontFamily: 'serif',
    fontWeight: '600',
  },
  paymentSummary: {
    color: '#3f51b5',
  },
  button: {
    textAlign: 'end',
    color: 'lightseagreen',
    fontWeight: '600',
  },
  chartButton: {
    color: '#ffffff',
    textTransform: 'initial',
    cursor: 'pointer',
  },
  divider: {
    marginTop: '30px',
  }
}));

const formatter = (amount) => {
  return (
    new Intl.NumberFormat('en-CA',
      { style: 'currency', currency: 'CAD' }
    ).format(amount)
  );
}

export default function MortgageSummary(props) {
  const classes = useStyles();
  const { values, amortization, goBack } = props;
  const [openChart, setOpenChart] = React.useState(false);

  const handleClickOpenChart = () => event => {
    event.preventDefault();
    setOpenChart(true);
  };

  const handleCloseChart = (prop) => {
    setOpenChart(false);
  };

  const handleForm = () => {
    goBack();
  }

  return (
    <div>
      <Grid container spacing={3}>
        <Grid item xs={6} className={classes.grid}>
          <Card className={classes.card}>
            <CardContent>
              <Typography variant="h5" className={classes.title}>{values.paymentFrequency} Payment</Typography>
              <Typography variant="h2" className={classes.amount}>
                {formatter(props.paymentAmount)}
              </Typography>
              <Button className={classes.chartButton} onClick={handleClickOpenChart()}><BarChart />View Payment Chart</Button>
            </CardContent>
          </Card>
        </Grid>
        <Grid item xs={6}>
          <Grid container spacing={3}>
            <Grid item xs={7}>
              <Typography variant="h5" className={classes.text}>Mortgage Details:</Typography>
            </Grid>
            <Grid item xs={5} className={classes.button}>
              <Button className={classes.button} onClick={handleForm}><Edit /> Edit Details</Button>
            </Grid>
            <Grid item xs={6}>
              <FormLabel component="legend" className={classes.formLabel}>Mortgage Amount: </FormLabel>
            </Grid>
            <Grid item xs={6}>
              {values.mortgageAmount}
            </Grid>
            <Grid item xs={6}>
              <FormLabel component="legend" className={classes.formLabel}>Amortization Period: </FormLabel>
            </Grid>
            <Grid item xs={6}>
              {(values.amortizationPeriod+' Years')}
            </Grid>
            <Grid item xs={6}>
              <FormLabel component="legend" className={classes.formLabel}>Interest Rate: </FormLabel>
            </Grid>
            <Grid item xs={6}>
              {values.interestRate+'%'}
            </Grid>
            <Grid item xs={6}>
              <FormLabel component="legend" className={classes.formLabel}>Interest Term: </FormLabel>
            </Grid>
            <Grid item xs={6}>
              {(values.interestTerm+' Years')}
            </Grid>
          </Grid>
        </Grid>
      </Grid>
      <Divider className={classes.divider} />
      <AmortizationTable amortization={amortization} />
      {openChart && <PaymentChart amortization={amortization} open={openChart} onClose={handleCloseChart} />}
    </div>
  );
}
