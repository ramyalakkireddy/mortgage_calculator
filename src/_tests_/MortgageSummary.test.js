import React from 'react';
import renderer from 'react-test-renderer';
import { mount, shallow } from 'enzyme';
import MortgageSummary from '../components/MortgageSummary';
import AmortizationTable from '../components/AmortizationTable';
import PaymentChart from '../components/PaymentChart';

const values = {
  mortgageAmount: '100000',
  amortizationPeriod: '25',
  paymentFrequency: 'Monthly',
  interestRate: '5',
  interestTerm: '5'
};
const paymentAmount = 581.61;
const amortization = [];

describe("MortgageSummary", () => {

  test('snapshot renders', () => {
    const component = shallow(<MortgageSummary values={values} paymentAmount={paymentAmount} amortization={amortization} />);
    let tree = component.getElements();
    expect(tree).toMatchSnapshot();
  });

  test("should render AmortizationTable component", () => {
    const component = shallow(<MortgageSummary values={values} paymentAmount={paymentAmount} amortization={amortization} />);
    let tree = component.getElements();
    expect(component.find(AmortizationTable).length).toEqual(1);
  })

});
