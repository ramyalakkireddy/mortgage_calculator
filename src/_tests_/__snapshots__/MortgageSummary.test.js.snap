// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MortgageSummary snapshot renders 1`] = `
Array [
  <div>
    <ForwardRef(WithStyles)
      container={true}
      spacing={3}
    >
      <ForwardRef(WithStyles)
        className="makeStyles-grid-5"
        item={true}
        xs={6}
      >
        <ForwardRef(WithStyles)
          className="makeStyles-card-3"
        >
          <ForwardRef(WithStyles)>
            <ForwardRef(WithStyles)
              className="makeStyles-title-2"
              variant="h5"
            >
              Monthly
               Payment
            </ForwardRef(WithStyles)>
            <ForwardRef(WithStyles)
              className="makeStyles-amount-4"
              variant="h2"
            >
              CA$581.61
            </ForwardRef(WithStyles)>
            <ForwardRef(WithStyles)
              className="makeStyles-chartButton-9"
              onClick={[Function]}
            >
              <UNDEFINED />
              View Payment Chart
            </ForwardRef(WithStyles)>
          </ForwardRef(WithStyles)>
        </ForwardRef(WithStyles)>
      </ForwardRef(WithStyles)>
      <ForwardRef(WithStyles)
        item={true}
        xs={6}
      >
        <ForwardRef(WithStyles)
          container={true}
          spacing={3}
        >
          <ForwardRef(WithStyles)
            item={true}
            xs={7}
          >
            <ForwardRef(WithStyles)
              className="makeStyles-text-1"
              variant="h5"
            >
              Mortgage Details:
            </ForwardRef(WithStyles)>
          </ForwardRef(WithStyles)>
          <ForwardRef(WithStyles)
            className="makeStyles-button-8"
            item={true}
            xs={5}
          >
            <ForwardRef(WithStyles)
              className="makeStyles-button-8"
              onClick={[Function]}
            >
              <UNDEFINED />
               Edit Details
            </ForwardRef(WithStyles)>
          </ForwardRef(WithStyles)>
          <ForwardRef(WithStyles)
            item={true}
            xs={6}
          >
            <ForwardRef(WithStyles)
              className="makeStyles-formLabel-6"
              component="legend"
            >
              Mortgage Amount: 
            </ForwardRef(WithStyles)>
          </ForwardRef(WithStyles)>
          <ForwardRef(WithStyles)
            item={true}
            xs={6}
          >
            100000
          </ForwardRef(WithStyles)>
          <ForwardRef(WithStyles)
            item={true}
            xs={6}
          >
            <ForwardRef(WithStyles)
              className="makeStyles-formLabel-6"
              component="legend"
            >
              Amortization Period: 
            </ForwardRef(WithStyles)>
          </ForwardRef(WithStyles)>
          <ForwardRef(WithStyles)
            item={true}
            xs={6}
          >
            25 Years
          </ForwardRef(WithStyles)>
          <ForwardRef(WithStyles)
            item={true}
            xs={6}
          >
            <ForwardRef(WithStyles)
              className="makeStyles-formLabel-6"
              component="legend"
            >
              Interest Rate: 
            </ForwardRef(WithStyles)>
          </ForwardRef(WithStyles)>
          <ForwardRef(WithStyles)
            item={true}
            xs={6}
          >
            5%
          </ForwardRef(WithStyles)>
          <ForwardRef(WithStyles)
            item={true}
            xs={6}
          >
            <ForwardRef(WithStyles)
              className="makeStyles-formLabel-6"
              component="legend"
            >
              Interest Term: 
            </ForwardRef(WithStyles)>
          </ForwardRef(WithStyles)>
          <ForwardRef(WithStyles)
            item={true}
            xs={6}
          >
            5 Years
          </ForwardRef(WithStyles)>
        </ForwardRef(WithStyles)>
      </ForwardRef(WithStyles)>
    </ForwardRef(WithStyles)>
    <ForwardRef(WithStyles)
      className="makeStyles-divider-10"
    />
    <AmortizationTable
      amortization={Array []}
    />
  </div>,
]
`;
